src/client/mod.rs

```diff
@@ -255,40 +252,36 @@ impl<C: hyper::client::Connect + Clone> Client<C> {
         }
     }
 
-    pub fn iter<L,T>(&self, gvr: &GroupVersionResource, namespace: Option<&str>, opts: ListOptions) -> Box<Stream<Item=T, Error=Error>>
-        where L: List<T> + DeserializeOwned + 'static, T: 'static
+    pub fn iter<L,T>(&self, gvr: &GroupVersionResource, namespace: Option<&str>, opts: ListOptions) -> Result<Box<Stream<Item=T, Error=Error>>, Error>
+        where L: List<T> + DeserializeOwned + 'static, T: 'static,
     {
-        let url = match self.url(gvr, namespace, None, opts.clone()) {
-            Ok(u) => Url::parse(&u.to_string()).unwrap(),
-            Err(e) => return Box::new(future::err(e).into_stream()),
-        };
+        let url = Url::parse(&self.url(gvr, namespace, None, opts.clone())?.to_string())?;
         let client_copy = self.clone();
-        let res = stream::unfold((url, opts, true), move |(mut url, mut opts, more)| {
-            if more {
+        let res = stream::unfold(Some((url, opts)), move |context| {
+            context.and_then(|(mut url, mut opts)| {
                 let req = hyper::Request::new(Method::Get, hyper_uri(url.clone()));
                 let res = client_copy.request(req)
                     .and_then(move |list: L| {
-                        let (opts, more) = {
+                        let next = {
                             let meta = list.listmeta();
                             debug!("listmeta: {:#?}", meta);
-                            let more = meta.continu.is_some();
-                            opts.continu = meta.continu.as_ref()
-                                .map(|s| s.clone())
-                                .unwrap_or_default();
-                            let query = serde_urlencoded::to_string(&opts)?;
-                            url.set_query(Some(&query));
-                            (opts, more)
+                            match meta.continu {
+                                Some(ref continu) => {
```

Comment:
I think I tried this during my experiments (see my repos other branches).  In fact yes - this is why I wanted to have a helper fn rather than a closure.

the query stringification can fail. We handle that with ?

If we use map, then the failure mode from that cannot be surfaced: we're not purely translating our continue closure here, we're also translating the status of the next page.
We could change Some(continue) to Ok(Some(continue)), and then we can express Err(the-serde-failure), as well as Ok(Some((url, opts))) for the next path out; but its getting a bit ridiculous vs this reasonably clear match().

Reply:
:+1:

Reply:
I did look at using Option::ok, but that maps wrongly.
We need something like this:
```
fn ok_maybe<I, O, E>(self: Option<I>, f: fn(i: I) -> Result<Option<O>, E>)
```

and maybe it can be nice. I will potter with it another time.